### PR TITLE
Bump JAX to 0.5.3

### PR DESCRIPTION
**Pull request by mudit2812**
This PR attempts to add support for JAX 0.5.3 in PennyLane.

[sc-84021] [sc-88693] [sc-90298]

### CODE REVIEW

File: tests/gradients/core/test_pulse_gradient.py [hunk 1]
```diff
@@ -1495,13 +1495,13 @@ def circuit(params):
...
-                for j, e in zip(j_shots, e_shots):
-                    assert qml.math.allclose(j[0], e, atol=tol, rtol=0.0)
+        if isinstance(shots, list):
+            for j_shots in jac:
+                for j, e in zip(j_shots, exp_jac):
+                    assert qml.math.allclose(j, e, atol=tol, rtol=0.0)
         else:
             for j, e in zip(jac, exp_jac):
-                assert qml.math.allclose(j[0], e, atol=tol, rtol=0.0)
+                assert qml.math.allclose(j, e, atol=tol, rtol=0.0)
```
Comment (mudit2812) [inline tests/gradients/core/test_pulse_gradient.py:1504]: This test wasn't checking the actual and expected jacobian correctly. It was passing in the past purely by coincidence. We just happened to be using a seed that worked with the tests. Another example of loose tolerances leading to false positives.

File: pennylane/devices/default_qubit.py [hunk 2]
```diff
@@ -1013,13 +1013,6 @@ def eval_jaxpr(
     def _backprop_jvp(self, jaxpr, args, tangents, execution_config=None):
         import jax
 
-        def _make_zero(tan, arg):
-            return (
-                jax.lax.zeros_like_array(arg) if isinstance(tan, jax.interpreters.ad.Zero) else tan
-            )
-
-        tangents = tuple(map(_make_zero, tangents, args))
```
Comment (mudit2812) [inline pennylane/devices/default_qubit.py:1021]: Moved this to `_capture_qnode`

File: pennylane/workflow/_capture_qnode.py [hunk 3]
```diff
@@ -350,6 +350,16 @@ def _finite_diff(args, tangents, **impl_kwargs):
 
 @debug_logger
 def _qnode_jvp(args, tangents, *, execution_config, device, qfunc_jaxpr, **impl_kwargs):
+
+    def _make_zero(tan, arg):
+        return (
+            jax.lax.zeros_like_array(arg).astype(tan.aval.dtype)
+            if isinstance(tan, jax.interpreters.ad.Zero)
+            else tan
+        )
```
Comment (mudit2812) [inline pennylane/workflow/_capture_qnode.py:359]: We need to use `astype` here because if the primals are integers, `tan.aval.dtype` will be `float0`, which is not a valid dtype to initialize `jnp.array`. The tangents _need_ to have the `float0` dtype for integer primals to work correctly with `jax.jvp`. With that in mind, it might not make sense to do the casting here and move it back to `default.qubit`, since `float0` is probably not a dtype that lightning can handle. Additionally, we may need to add this validation to the `out_tangents` returned by `Device.jaxpr_jvp`.

File: .github/workflows/unit-test.yml [hunk 4]
```diff
@@ -189,7 +189,7 @@ jobs:
         id: pytest_args
         env:
           PYTEST_COVERAGE_ARGS: ${{ inputs.pytest_coverage_flags }}
-          PYTEST_PARALLELISE_ARGS: -n auto
+          PYTEST_PARALLELISE_ARGS: ''
```
Comment (albi3ro) [inline .github/workflows/unit-test.yml:192]:
```suggestion
          PYTEST_PARALLELISE_ARGS: -n auto
```

File: tests/devices/default_qubit/test_default_qubit.py [hunk 5]
```diff
@@ -1486,8 +1490,10 @@ def test_finite_shots_postselection_defer_measurements(self, max_workers):
 
         # Executing with different calls to dev.execute should give the same results
         res = [dev.execute(tape, config) for tape in tapes]
+        print(res)
```
Comment (albi3ro) [inline tests/devices/default_qubit/test_default_qubit.py:1493]:
```suggestion
```

File: tests/devices/default_qubit/test_default_qubit.py [hunk 6]
```diff
@@ -1486,8 +1490,10 @@ def test_finite_shots_postselection_defer_measurements(self, max_workers):
...
         shapes = [qml.math.shape(r) for r in res]
         assert len(shapes) == n_tapes
+        print(shapes)
```
Comment (albi3ro) [inline tests/devices/default_qubit/test_default_qubit.py:1496]:
```suggestion
```

File: .github/workflows/interface-unit-tests.yml [hunk 7]
```diff
@@ -494,7 +494,7 @@ jobs:
       additional_pip_packages: |
         pyzx matplotlib stim quimb mitiq ply optax scipy-openblas32>=0.3.26 qualtran
         git+https://github.com/PennyLaneAI/pennylane-qiskit.git@master
-        ${{ needs.default-dependency-versions.outputs.jax-version }}
+        jax==0.4.28 jaxlib==0.4.28
```
Comment (mudit2812) [inline .github/workflows/interface-unit-tests.yml:497]: We should probably wait for Catalyst to merge their upgrade PR before we merge this one, and revert this file accordingly. What do you think?

```suggestion
        ${{ needs.default-dependency-versions.outputs.jax-version }}
```

File: pennylane/capture/custom_primitives.py [hunk 8]
```diff
@@ -17,7 +17,8 @@
 from enum import Enum
 from typing import Union
 
-import jax
+# import jax
```
Comment (mudit2812) [inline pennylane/capture/custom_primitives.py:20]:
```suggestion
```

File: pennylane/devices/qubit/simulate.py [hunk 9]
```diff
@@ -322,6 +322,7 @@ def simulate(
     tensor([0.68117888, 0.        , 0.31882112, 0.        ], requires_grad=True))
 
     """
+    circuit = circuit.copy()
```
Comment (mudit2812) [inline pennylane/devices/qubit/simulate.py:325]: Why was this added?

File: pennylane/workflow/resolution.py [hunk 10]
```diff
@@ -52,6 +52,29 @@ def _get_jax_interface_name() -> Interface:
...
+            f"PennyLane is currently not compatible with versions of less than {min_jax_version}. "
+            f"You have version {jax_version} installed.",
+            RuntimeWarning,
+        )
+    if Version(max_jax_version) < Version(jax_version):  # pragma: no cover
+        warn(
+            f"PennyLane is currently not compatible with versions of greater than {max_jax_version}. "
+            f"You have version {jax_version} installed.",
+            RuntimeWarning,
+        )
```
Comment (mudit2812) [inline pennylane/workflow/resolution.py:75]: So we're essentially updating this warning so that it will only get raised if a user has jax >= 1.0 installed.

File: pennylane/devices/qubit/simulate.py (hunk 9 above)
Comment (albi3ro) [reply pennylane/devices/qubit/simulate.py:325]: When debugging `test_default_qubit.py:test_finite_shots_postselection_defer_measurements`. Each circuit executed with about half the number of shots as the previous one, since `simulate` modified its input.

File: pennylane/workflow/resolution.py (hunk 10 above)
Comment (albi3ro) [reply pennylane/workflow/resolution.py:75]: The idea is that we keep the instructure for warning about jax version around, so it's easy to set upper and lower limits whenever we need to.

File: pennylane/capture/explanations.md [hunk 11]
```diff
@@ -428,6 +426,9 @@ Now in our jaxpr, we can see thet `PrimitiveClass2` returns something of type `A
 
 # Non-interpreted primitives
 
+**WARNING:** THIS EXPLANATION IS SPECIFIC TO JAX 0.4.28 AND NO LONGER APPLIES
```
Comment (PietropaoloFrisoni) [inline pennylane/capture/explanations.md:429]: We might as well remove this explanation file at this point, but if you prefer to keep it, I have no objections.

File: pennylane/workflow/resolution.py (hunk 10 above)
Comment (PietropaoloFrisoni) [reply pennylane/workflow/resolution.py:75]: Is this warning still needed after the bugfix release PR?

File: tests/capture/transforms/test_mcm_execution.py [hunk 12]
```diff
@@ -268,6 +268,7 @@ def f():
         res = tuple(
             dev.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, execution_config=config)[0] for _ in range(10)
         )
+        print(res)
```
Comment (PietropaoloFrisoni) [inline tests/capture/transforms/test_mcm_execution.py:271]:
```suggestion
```

File: tests/capture/workflow/test_capture_qnode.py [hunk 13]
```diff
@@ -1881,6 +1883,7 @@ def circuit(x):
         assert spy.call_count > 1
 
     # pylint: disable=unused-argument
+    @pytest.mark.xfail  # think v0.5.3 broke dynamic shapes and static argnums
```
Comment (PietropaoloFrisoni) [inline tests/capture/workflow/test_capture_qnode.py:1886]:
```suggestion
    @pytest.mark.xfail  # think JAX 0.5.3 broke dynamic shapes and static argnums
```

for future reference : )

File: tests/capture/workflow/test_capture_qnode.py [hunk 14]
```diff
@@ -1923,6 +1926,7 @@ def circuit(x, y):
...
```
Comment (PietropaoloFrisoni) [inline tests/capture/workflow/test_capture_qnode.py:1929]:
```suggestion
    @pytest.mark.xfail  # think JAX 0.5.3 broke dynamic shapes and static argnums
```

File: tests/devices/default_qubit/test_default_qubit.py [hunk 15]
```diff
@@ -1462,18 +1466,18 @@ def test_different_executions_same_prng_key(self, max_workers):
 
         assert np.all(result1 == result2)
 
-    @pytest.mark.parametrize("max_workers", max_workers_list)
-    def test_finite_shots_postselection_defer_measurements(self, max_workers):
+    # @pytest.mark.parametrize("max_workers", max_workers_list)
```
Comment (PietropaoloFrisoni) [inline tests/devices/default_qubit/test_default_qubit.py:1469]: Should we delete this `pytest` directive completely?

File: tests/devices/default_qubit/test_default_qubit.py [hunk 16]
```diff
@@ -1462,18 +1466,18 @@ def test_different_executions_same_prng_key(self, max_workers):
...
         multiple times."""
         import jax
 
-        dev = qml.device("default.qubit", max_workers=max_workers, seed=jax.random.PRNGKey(678))
+        dev = qml.device("default.qubit", seed=jax.random.PRNGKey(234))
 
         mv = qml.measure(0, postselect=1)
         qs = qml.tape.QuantumScript(
-            [qml.Hadamard(0), mv.measurements[0]], [qml.sample(wires=0)], shots=100
+            [qml.Hadamard(0), mv.measurements[0]], [qml.sample(wires=0)], shots=500
```
Comment (PietropaoloFrisoni) [inline tests/devices/default_qubit/test_default_qubit.py:1480]: Did you increment the number of shots because of precision issues?

**Review by PietropaoloFrisoni — COMMENTED** [review]
Looks great. I am a little confused by some changes, but I'll re-read the Notion doc about JAX updates

File: pennylane/capture/explanations.md (hunk 11 above)
Comment (albi3ro) [reply pennylane/capture/explanations.md:429]: I think i'll keep in around in case we want to get back to this.

File: pennylane/workflow/resolution.py (hunk 10 above)
Comment (albi3ro) [reply pennylane/workflow/resolution.py:75]: I'm betting that we will need this code again next time jax releases.  Though might be a good idea to figure out the min version now too.

File: tests/devices/default_qubit/test_default_qubit.py (hunk 15 above)
Comment (albi3ro) [reply tests/devices/default_qubit/test_default_qubit.py:1469]: Potetentially worth bringing back once we figure out jax with multiprocessing.

File: tests/devices/default_qubit/test_default_qubit.py (hunk 16 above)
Comment (albi3ro) [reply tests/devices/default_qubit/test_default_qubit.py:1480]: yep. The numbers from the key changed due to a different version, so we needed more shots.

**Review by albi3ro — APPROVED** [review]
Given both @mudit2812 and I have worked on this PR, but we both approve, I'll give one approval from both of us.

**Review by PietropaoloFrisoni — APPROVED** [review]

File: pennylane/capture/base_interpreter.py [hunk 17]
```diff
@@ -430,12 +432,14 @@ def _(self, x, *dyn_shape, shape, broadcast_dimensions):
     # needs custom primitive as jax.core.eval_jaxpr will error out with this
     new_shape = _fill_in_shape_with_dyn_shape(dyn_shape, shape)
 
-    return jax.lax.broadcast_in_dim(x, new_shape, broadcast_dimensions=broadcast_dimensions)
+    return jax.lax.broadcast_in_dim(
+        x, new_shape, broadcast_dimensions=broadcast_dimensions, out_sharding=sharding
```
Comment (paul0403) [inline pennylane/capture/base_interpreter.py:436]: Thanks! This new `out_sharding` kwarg was causing trouble for Catalyst. Glad to see it's resolved here : )

File: pennylane/ops/functions/assert_valid.py [hunk 18]
```diff
@@ -327,8 +327,12 @@ def _check_capture(op):
...
     qml.capture.enable()
     try:
-        jaxpr = jax.make_jaxpr(lambda obj: obj)(op)
-        data, _ = jax.tree_util.tree_flatten(op)
+        data, struct = jax.tree_util.tree_flatten(op)
+
+        def test_fn(*args):
+            return jax.tree_util.tree_unflatten(struct, args)
+
+        jaxpr = jax.make_jaxpr(test_fn)(*data)
```
Comment (paul0403) [inline pennylane/ops/functions/assert_valid.py:335]: Just a warning from me, I think the `make_jaxpr` functions got a new kwarg `debug_info` in 0.5.1 that will become mandatory in 0.6.0. https://github.com/jax-ml/jax/issues/26480

This is how I dealt with it in Catalyst https://github.com/PennyLaneAI/catalyst/pull/1681. 

Not sure about whether it's needed in this particular one though. Maybe this one does not need it.

Comment (mudit2812) [reply pennylane/ops/functions/assert_valid.py:335]: `jax.make_jaxpr` has not changed. The usage of internal data structures (which I don't think we interact with) are the ones that need updating.
